extract_feature_region.py

```diff
@@ -0,0 +1,156 @@
+dataset_to_folder_mapper["coco_2015_test"] = "test2015"
+
+
+def extract_feature_argument_parser():
+    parser = argparse.ArgumentParser(description="Region feature extraction")
+    parser.add_argument(
+        "--config-file", default="", metavar="FILE", help="path to config file"
+    )
+    parser.add_argument(
+        "--dataset",
+        help="name of the dataset",
+        default="coco_2014_train",
+        choices=["coco_2014_train", "coco_2014_val", "coco_2015_test", "visual_genome"],
+    )
+    parser.add_argument(
+        "--dataset-path",
+        help="path to image folder dataset, if not provided expects a detection dataset",
+        default="",
+    )
+    parser.add_argument(
+        "opts",
+        help="Modify config options using the command-line",
+        default=None,
+        nargs=argparse.REMAINDER,
+    )
+    return parser
+
+
+def extract_feature_on_dataset(model, data_loader, dump_folder):
+    for idx, inputs in enumerate(tqdm.tqdm(data_loader)):
+        with torch.no_grad():
+            image_id = inputs[0]["image_id"]
+
+            # compute features and proposals
+            images = model.preprocess_image(inputs)
+            features = model.backbone(images.tensor)
+            proposals, _ = model.proposal_generator(images, features)
+
+            # pooled features and box predictions
+            box_features, pooled_features = model.roi_heads.get_rois_features(
+                features, proposals
+            )
+            predictions = model.roi_heads.box_predictor(pooled_features)
+            cls_probs = F.softmax(predictions[0], dim=-1)
+            cls_probs = cls_probs[:, :-1]  # background is last
+            predictions, r_indices = model.roi_heads.box_predictor.inference(
+                predictions, proposals
+            )
+
+            # postprocess
+            height = inputs[0].get("height")
+            width = inputs[0].get("width")
+            r = postprocessing.detector_postprocess(predictions[0], height, width)
+
+            bboxes = r.get("pred_boxes").tensor
+            classes = r.get("pred_classes")
+            cls_probs = cls_probs[r_indices]
+            pooled_features = pooled_features[r_indices]
+
+            assert (
+                bboxes.size(0)
+                == classes.size(0)
+                == cls_probs.size(0)
+                == pooled_features.size(0)
+            )
+
+            # save info and features
+            info = {
+                "bbox": bboxes.cpu().numpy(),
+                "num_boxes": bboxes.size(0),
+                "objects": classes.cpu().numpy(),
+                "image_height": r.image_size[0],
+                "image_width": r.image_size[1],
+                "cls_prob": cls_probs.cpu().numpy(),
+            }
+            features = pooled_features.cpu().numpy()
+            np.save(os.path.join(dump_folder, str(image_id) + "_info.npy"), info)
```

Comment:
We are saving in this manner as this is the old format of input in Pythia (mmf). I have no strong preference here and I can keep it to one file. However we will need another extraction script that we can keep in mmf repository in that case.

Reply:
Since you said old, does it mean in the new format you guys are considering to merge it into one file? I wonder the advantage of having separate files -- those meta information, if not saved with each individual feature, should be saved in a single file (since the size is small) like .json/.npy to save flooding inodes.

Reply:
I have merged these to a single file.

Reply:
Thanks! If you have time, please fix other comments -- especially the ones of extracting region features exactly following the original BUTD procedure.